kong/plugins/datadog/statsd_logger.lua

```diff
@@ -0,0 +1,86 @@
+local setmetatable = setmetatable
+local ngx_socket_udp = ngx.socket.udp
+local ngx_log = ngx.log
+local table_concat = table.concat
+
+local statsd_mt = {}
+statsd_mt.__index = statsd_mt
+
+function statsd_mt:new(conf)
+
+  local sock = ngx_socket_udp()
+  sock:settimeout(conf.timeout)
+  local _, err = sock:setpeername(conf.host, conf.port)
+  if err then
+     ngx_log(ngx.ERR, "failed to connect to "..conf.host..":"..tostring(conf.port)..": ", err)
+    return
+  end
+  
+  local statsd = {
+    host = conf.host,
+    port = conf.port,
+    socket = sock,
+  }
+  return setmetatable(statsd, statsd_mt)
+end
+
+function statsd_mt:create_statsd_message(stat, delta, kind, sample_rate)
+  local rate = ""
+  if sample_rate and sample_rate ~= 1 then 
+    rate = "|@"..sample_rate 
+  end
+  
+  local message = {
+    "kong",
+    ".",
+    stat,
+    ":",
+    delta,
+    "|",
+    kind,
+    rate
+  }
+  return table_concat(message, "")
+end
+
+function statsd_mt:close_socket()
+  local ok, err = self.socket:close()
+  if not ok then
+    ngx_log(ngx.ERR, "failed to close connection from "..self.host..":"..tostring(self.port)..": ", err)
```

Comment:
The same could be said for all `ngx.ERR` but it's less likely that they'll be called as often and have any kind of impact. Consider caching them just for consistency but if not it's fine.